activerecord/lib/active_record/connection_adapters/abstract/transaction.rb

```diff
@@ -317,25 +317,31 @@ def within_new_transaction(isolation: nil, joinable: true)
           end
           raise
         ensure
-          if !error && transaction
-            if Thread.current.status == "aborting"
-              rollback_transaction
+          if transaction
+            if error
+              # @connection still holds an open transaction, so we must not
+              # put it back in the pool for reuse
+              @connection.disconnect!  unless transaction.state.rolledback?
```

Comment:
`#discard!`'s documentation says:

> Prevent `@connection`'s finalizer from touching the socket, or otherwise communicating with its server, when it is collected.
>
> Unlike disconnect!, this will not communicate with the server.

Original thinking here was that we don't want to risk not being able to talk to the server _again_ and having this call also 💥. I'm not entirely confident that `#discard!` does everything we want though.

Reply:
Using `discard!` here will just make this connection be in an undefined state. In other words, a thread can still checkout this connection and any method called on it will raise an exception. What we want is to remove the connection from the connection poll and after that discard it.